File: DataStructure/BinarySearch/binary_search.py

'''
Author: lanyongliang
Email: lanyongliang@xdf.cn
Date: 2021-07-15 00:12:02
LastEditors: Please set LastEditors
LastEditTime: 2021-07-15 01:21:58
FilePath: \BinarySearch\binary_search.py
'''

import random

class BinarySearch(object):

    """
    1、二分查找必须是有序类表
    2、选好起始查找位置，中间开始，减少一半的查找元素
    3、时间复杂度为 ${log_2 10}$
    
    """
    def __call__(self, search_list, num):
        output = self.search(search_list, num)
        return output

    def search(self, search_list, num):

        end = len(search_list)
        start = 0
        guess = (start + end) // 2
        
        while start < guess:
            if len(search_list) >= 1:
                if search_list[guess] == num:
                    return guess
                elif search_list[guess] > num:
                    end = guess
                else:
                    start = guess + 1
                guess = (start + end) // 2
            else:
                return None


random_list = random.sample(range(0,100),20)
random_list.sort()
binary_search = BinarySearch()
print("查找集合：{}".format(random_list))

output = binary_search(random_list, 10)
print("output:{}".format(output))

File: DataStructure/BinarySearch/binarySearch.cpp
/*
 * @Author: lanyongliang
 * @Email: lanyongliang@xdf.cn
 * @Date: 2021-07-14 15:25:56
 * @LastEditors: Please set LastEditors
 * @LastEditTime: 2021-07-16 00:45:46
 * @FilePath: \AlgorithmPractice\DataStructure\BinarySearch\binarySearch.cpp
 */


#include <stdio.h>
#include <windows.h>
#include <iostream>
#include <math.h>
using namespace std;



void random(int * arr, int start, int end, int count){
    srand(2);
    for (int i=0; i<count; i++){
        arr[i] = rand() % (end - start) + start;
    }

}
int search(int searchArr[], int num, int length){

    int start = 0;
    int end = length;
    std::cout<<end<<std::endl;
    int guess = floor((start + end) / 2);

    while (start <= guess){
        if (searchArr[guess] == num){
            return guess;
        }
        else if(searchArr[guess] < num){
            start = guess + 1;
        }
        else {


            end = guess;
        }
        guess = floor((start + end) /2 );

    }
    return -1;

}


int  main()
{
    int arr[10] = {10,20,24,28,29,30,50,80,99,100};
    int index;
    // std::cout<<sizeof(arr)<<"\n"<<std::endl;
    // random(arr,10,30,10);
    // for (int i=0; i<10; i++){
    //     std::cout<<"i: " <<arr[i]<<std::endl;
    // }

    index = search(arr, 28, 10);
    std::cout<<arr<<"\n"<<index<<std::endl;
    return index;
}